java/serve/src/main/java/io/ray/serve/ReplicaSet.java

```diff
@@ -0,0 +1,153 @@
+      actorNames.forEach(
+          name ->
+              workerReplicas.add((ActorHandle<RayServeWrappedReplica>) Ray.getActor(name).get()));
+    }
+
+    Set<ActorHandle<RayServeWrappedReplica>> added =
+        new HashSet<>(Sets.difference(workerReplicas, inFlightQueries.keySet()));
+    Set<ActorHandle<RayServeWrappedReplica>> removed =
+        new HashSet<>(Sets.difference(inFlightQueries.keySet(), workerReplicas));
+
+    added.forEach(actorHandle -> inFlightQueries.put(actorHandle, Sets.newConcurrentHashSet()));
+    removed.forEach(actorHandle -> inFlightQueries.remove(actorHandle));
+
+    if (added.size() > 0 || removed.size() > 0) {
+      LOGGER.info("ReplicaSet: +{}, -{} replicas.", added.size(), removed.size());
+    }
+  }
+
+  /**
+   * Given a query, submit it to a replica and return the object ref. This method will keep track of
+   * the in flight queries for each replicas and only send a query to available replicas (determined
+   * by the backend max_concurrent_quries value.)
+   *
+   * @param query the incoming query.
+   * @return ray.ObjectRef
+   */
+  public ObjectRef<Object> assignReplica(Query query) {
+    String endpoint = query.getMetadata().getEndpoint();
+    numQueuedQueries.incrementAndGet();
+    RayServeMetrics.execute(
+        () ->
+            numQueuedQueriesGauge.update(
+                numQueuedQueries.get(),
+                TagKey.tagsFromMap(ImmutableMap.of(RayServeMetrics.TAG_ENDPOINT, endpoint))));
+    ObjectRef<Object> assignedRef =
+        tryAssignReplica(query); // TODO controll concurrency using maxConcurrentQueries
+    numQueuedQueries.decrementAndGet();
+    RayServeMetrics.execute(
+        () ->
+            numQueuedQueriesGauge.update(
+                numQueuedQueries.get(),
+                TagKey.tagsFromMap(ImmutableMap.of(RayServeMetrics.TAG_ENDPOINT, endpoint))));
+    return assignedRef;
+  }
+
+  /**
+   * Try to assign query to a replica, return the object ref if succeeded or return None if it can't
+   * assign this query to any replicas.
+   *
+   * @param query query the incoming query.
+   * @return ray.ObjectRef
+   */
+  private ObjectRef<Object> tryAssignReplica(Query query) {
+
+    List<ActorHandle<RayServeWrappedReplica>> handles = new ArrayList<>(inFlightQueries.keySet());
+    if (CollectionUtil.isEmpty(handles)) {
+      throw new RayServeException("ReplicaSet found no replica.");
+    }
+    int randomIndex = RandomUtils.nextInt(0, handles.size());
+    ActorHandle<RayServeWrappedReplica> replica =
+        handles.get(randomIndex); // TODO controll concurrency using maxConcurrentQueries
+    LOGGER.debug("Assigned query {} to replica {}.", query.getMetadata().getRequestId(), replica);
+    return replica
+        .task(RayServeWrappedReplica::handleRequest, query.getMetadata(), query.getArgs())
+        .remote();
+  }
+
+  public synchronized int
+      drainCompletedObjectRefs() { // TODO when using maxConcurrentQueries, change it to private.
```

Comment:
remove this or comment out because it's never called

Reply:
OK.